### PR TITLE
AL09: Handle case sensitivity with aliases

### LINKED ISSUE #5636: AL09 - Case sensitivity with unquoted identifiers: TSQL
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

When using sql fluff fix it rewrites our SQL queries and removes "self aliasing" even though the casing is different and even though this is nescessary.

### Expected Behaviour

I do not want "ID" to be interpreted as equal to "id" or "Id".

### Observed Behaviour

The following is changed by sqlfluff fix:
```
SELECT
	id AS  ID,
	date AS Date,
	description AS Description,
FROM
[...]
```
into:
```
SELECT
	id,
	date,
	description,
FROM
[...]
```

### How to reproduce

Use the above SQL query and run sqlfluff lint or fix on it

### Dialect

tsql

### Version

  - repo: https://github.com/sqlfluff/sqlfluff
    rev: 3.0.0a5
    hooks:
      - id: sqlfluff-fix
      - id: sqlfluff-lint

### Configuration

[tool.sqlfluff.core]
large_file_skip_byte_limit = 100000
dialect = "tsql"
ignore = "templating"

[tool.sqlfluff.indentation]
tab_space_size = "4"
indent_unit = "tab"

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### LINKED ISSUE #6019: AL09 - Case sensitivity with unquoted identifiers: Clickhouse
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

The fix command unexpectedly deletes aliases when ignoring capitalization rules for identifiers.

Submitted query:
SELECT
    myidcol AS col_id,
    my_expected_drop_alias AS my_expected_drop_alias,
    My_Wrong_Case_Col as my_wrong_case_col,-- noqa: CP02
    my_good_col,
FROM
    file('myfile.parquet', 'Parquet')

Reformatted query:
SELECT
    myidcol AS col_id,
    my_expected_drop_alias,
    My_Wrong_Case_Col,-- noqa: CP02
    my_good_col,
FROM
    file('myfile.parquet', 'Parquet')

### Expected Behaviour

Expected result:
SELECT
    myidcol AS col_id,
    my_expected_drop_alias,
    My_Wrong_Case_Col as my_wrong_case_col,-- noqa: CP02
    my_good_col,
FROM
    file('myfile.parquet', 'Parquet')

### Observed Behaviour

When AL09 is used, but CP02 is ignored, AL09 is incorrectly triggered.

### How to reproduce

Example queries provided

### Dialect

Clickhouse

### Version

sqlfluff, version 3.1.0

### Configuration

Default config + inline noqa flag.

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by keraion**
### Brief summary of the change made

This handles a set of situations:
- Retains the usage of an alias if the case of the alias is different from the column name referenced. (e.g. `"COL" as "col"`)
- If the column name is case-insensitive for access, and the alias is a different case than the column name, replace the column name with the case used by the alias. (e.g. `col1 as COL` would be replace with `COL1`) This is dialect specific.
- fixes #5636
- fixes #6019

### Are there any other side effects of this change that we should be aware of?
None, but additional dialects that support the result set with case-insensitive column names will need to be added.

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: test/fixtures/rules/std_rule_cases/AL09.yml [hunk 1]
```diff
@@ -81,13 +90,18 @@ test_fail_self_alias_quoted:
             col_b as new_col_b
         from foo
 
-test_pass_self_alias_case_insensitive:
+test_fail_self_alias_case_sensitive_column:
```
Comment (aayushr7) [inline test/fixtures/rules/std_rule_cases/AL09.yml:93]: This test should pass as `"col_a"` is a Literal.

**Review by aayushr7 — CHANGES_REQUESTED** [review]

**Comment by aayushr7** [comment]
I have created a PR https://github.com/sqlfluff/sqlfluff/pull/6026 to handle case changes using unquoted identifiers and apply the relevant fix.

File: src/sqlfluff/rules/aliasing/AL09.py [hunk 2]
```diff
@@ -76,9 +80,7 @@ def _eval(self, context: RuleContext) -> EvalResultType:
                 # If column has either a naked_identifier or quoted_identifier
                 # (not positional identifier like $n in snowflake)
                 # then continue
-                if column.get_child("naked_identifier") or column.get_child(
-                    "quoted_identifier"
-                ):
+                if column.get_child("naked_identifier", "quoted_identifier"):
```
Comment (alanmcruickshank) [inline src/sqlfluff/rules/aliasing/AL09.py:83]: Nice 👌

File: src/sqlfluff/rules/aliasing/AL09.py [hunk 3]
```diff
@@ -60,11 +60,15 @@ def _eval(self, context: RuleContext) -> EvalResultType:
...
 
-        for clause_element in children.select(sp.is_type("select_clause_element")):
-            clause_element_raw_segments = (
-                clause_element.get_raw_segments()
-            )  # col_a as col_a
+        # The dialect supports case insensitive naming for a column name, (e.g. COL1
+        # is the same as col1) but the value in the result set name is case sensitive.
+        # We use this to replace the column reference with the alias instead of just
+        # removing the alias.
+        dialect_supports_ci_names_cs_result = ["tsql", "bigquery"]
```
Comment (alanmcruickshank) [inline src/sqlfluff/rules/aliasing/AL09.py:67]: I think this comment is hard to follow. Perhaps this would be clearer:
```suggestion
        # Some dialects are case insensitive when referencing columns (e.g. COL1 is
        # the same as col1), but the case of the reference does follow through into
        # the result set (i.e. where COL1 and col1 refer to the same column, the column
        # label in the result set will be different). In these cases, if the alias and
        # the column compare equal (case insensitive), but have a different case, we
        # replace the reference with the alias to achieve a simpler but consistent
        # result.
        dialect_supports_ci_names_cs_result = ["tsql", "bigquery"]
```

File: src/sqlfluff/rules/aliasing/AL09.py [hunk 4]
```diff
@@ -93,21 +95,29 @@ def _eval(self, context: RuleContext) -> EvalResultType:
...
+                    )
 
                     assert whitespace and column_identifier and alias_identifier
 
                     # Column self-aliased
-                    if column_identifier.raw_upper == alias_identifier.raw_upper:
+                    if (
+                        column_identifier.raw_normalized()
+                        # We casefold the alias only if it relevant in the result set.
+                        == alias_identifier.raw_normalized(ci_names_cs_result)
```
Comment (alanmcruickshank) [inline src/sqlfluff/rules/aliasing/AL09.py:107]:
```suggestion
                        # We casefold the alias only for relevant dialects.
                        == alias_identifier.raw_normalized(casefold=ci_names_cs_result)
```

**Review by alanmcruickshank — CHANGES_REQUESTED** [review]
Nice. Logic checks out 👍 .

Took me a while to work out what was going on, and as a result I think we could make the commentary clearer for the next person that comes along. I've suggested some improvements.

**Comment by aayushr7** [comment]
Hi @alanmcruickshank , 

I don't agree that the logic is correct. 

For example in this case: 
```
test_fail_self_alias_case_sensitive_column:
    fail_str:
        select
            "col_a" as col_a,
            col_b as new_col_b
        from foo
    fix_str:
        select
            "col_a",
            col_b as new_col_b
        from foo
```

The original query is actually doing a Case Change, from "col_a" -> "COL_A" as unquoted identifiers result in uppercase column names by default depending on the configuration. See example below:

Screenshot 2024-08-10 at 09 29 26
 
Therefore, it is a valid case change and should not be considered re-alias of the column. 

Could you please have a look at the PR that I shared in the comment? https://github.com/sqlfluff/sqlfluff/pull/6026

**Comment by bitomukesh** [comment]
AI Code Review is in progress (usually takes <3 minutes unless it's a very large PR).

File: src/sqlfluff/rules/aliasing/AL09.py [hunk 5]
```diff
@@ -60,11 +60,18 @@ def _eval(self, context: RuleContext) -> EvalResultType:
...
-                clause_element.get_raw_segments()
-            )  # col_a as col_a
+        # Some dialects are case insensitive when referencing columns (e.g. COL1 is
+        # the same as col1), but the case of the reference does follow through into
+        # the result set (i.e. where COL1 and col1 refer to the same column, the column
+        # label in the result set will be different). In these cases, if the alias and
+        # the column compare equal (case insensitive), but have a different case, we
+        # replace the reference with the alias to achieve a simpler but consistent
+        # result.
+        dialect_supports_ci_names_cs_result = ["tsql", "bigquery"]
```
Comment (bitomukesh) [inline src/sqlfluff/rules/aliasing/AL09.py:70]: <div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/sqlfluff/sqlfluff/pull/6022#issuecomment-2296942979>#0cc38e</a><span> - 08/19/2024, 04:14 pm</span>
<br>

<div><small>🟠</small> Medium importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> The list 'dialect_supports_ci_names_cs_result' should be defined as a constant since it is not meant to be modified.
</div>
<div id="fix">
<b>Fix:</b> Define 'dialect_supports_ci_names_cs_result' as a constant by using all capital letters and underscores.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -69,7 +69,7 @@
      # result.
 -    dialect_supports_ci_names_cs_result = ["tsql", "bigquery"]
 +    DIALECT_SUPPORTS_CI_NAMES_CS_RESULT = ["tsql", "bigquery"]

      ci_names_cs_result = context.dialect.name in dialect_supports_ci_names_cs_result
```
</div>
</details>
</div>
<br>


</div>

Is this a valid issue, or was it incorrectly flagged by the Agent?
- [ ] it was incorrectly flagged

File: test/fixtures/rules/std_rule_cases/AL09.yml (hunk 1 above)
Comment (bitomukesh) [inline test/fixtures/rules/std_rule_cases/AL09.yml:93]: <div>

<strong>Bito Code Review Agent Run</strong> <a href=https://github.com/sqlfluff/sqlfluff/pull/6022#issuecomment-2296942979>#0cc38e</a><span> - 08/19/2024, 04:14 pm</span>
<br>

<div><small>🔴</small> High importance</div>

<div id="suggestion">
<div id="issue">
<b>Issue:</b> The test case 'test_fail_self_alias_case_sensitive_column' is renamed from 'test_pass_self_alias_case_insensitive', but the description and the logic of the test case are not updated to reflect the new purpose. This can lead to confusion and incorrect test results.
</div>
<div id="fix">
<b>Fix:</b> Update the description and logic of the test case to accurately reflect the new purpose of testing case-sensitive column aliases.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -84,13 +93,18 @@
  test_fail_self_alias_case_sensitive_column:
      # Renaming the column name from case-sensitive to case-insensitive
 -    pass_str:
 +    fail_str:
          select
              "col_a" as col_a,
              col_b as new_col_b
          from foo
 +    fix_str:
 +        select
 +            "col_a",
 +            col_b as new_col_b
 +        from foo
```
</div>
</details>
</div>
<br>


</div>

Is this a valid issue, or was it incorrectly flagged by the Agent?
- [ ] it was incorrectly flagged

**Comment by alanmcruickshank** [comment]
Closing this as superseded by #6333.

**Comment by alanmcruickshank** [comment]
Closing this as superseded by #6333.
